### PR TITLE
Refactors

### PR DESCRIPTION
**Pull request by angrybayblade**
### **PR Type**
Enhancement

___

### **Description**
- Defined `Composio` SDK client and various model abstractions for server objects.
- Implemented HTTP client for both synchronous and asynchronous requests.
- Defined CLI context, entry point, and commands for the `composio` CLI tool.
- Added custom exceptions for HTTP and client errors.
- Defined global constants for environment variables and local cache.
- Implemented local storage helpers and user data storage class.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Define CLI group and add commands.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/cli/__init__.py

<li>Added CLI group definition for <code>composio</code>.<br> <li> Imported and added commands for login, apps, triggers, and <br>integrations.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-8673a267670e10eac352c3a1cf5fcdbb1dd3c8f1ed12ebb9b5126b1836d45e36">+23/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>__main__.py</strong><dd><code>Define entry point for CLI tool.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/cli/__main__.py

- Added entry point for the CLI tool.

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-b514975d1f64957d61b896aeda884b9bc899203eab50bf50c0f12519b5f7f49a">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>context.py</strong><dd><code>Define CLI context and helper decorator.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/cli/context.py

<li>Defined CLI context for runtime.<br> <li> Added properties for console, cache directory, user data, and client.<br> <li> Implemented <code>pass_context</code> decorator.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-8c73b317a8368ff08248ca667e900437d2304fcb0de128a930cfe7583d301bfb">+89/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Define SDK client and model abstractions.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/client/__init__.py

<li>Defined <code>Composio</code> SDK client.<br> <li> Added <code>Collection</code> class for server objects.<br> <li> Implemented various models and collections (e.g., <code>ConnectedAccount</code>, <br><code>App</code>, <code>Trigger</code>, <code>Action</code>).<br> <li> Added methods for client authentication and session validation.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-17f4a7383ae2819192cd4556b82eeb2389ca39bd3f6919573c0d8c2edd61e38f">+510/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>endpoints.py</strong><dd><code>Define HTTP endpoint abstraction.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/client/endpoints.py

<li>Defined HTTP endpoint abstraction.<br> <li> Added nested endpoint classes for CLI, apps, actions, and triggers.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-c909ba684fac5c88964d53da1ef4b621c04292a9745d6d29352de5152d35f21f">+102/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>exceptions.py</strong><dd><code>Define custom exceptions for HTTP and client errors.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/client/exceptions.py

- Defined custom exceptions for HTTP errors and client errors.

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-730cec88a9956de9ada21fd598e87ce172bc8c4c895c8ec3be217c7f1051796f">+37/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>http.py</strong><dd><code>Implement HTTP client for SDK.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/client/http.py

- Implemented HTTP client for synchronous and asynchronous requests.

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-89f1a3e88fb15b876e09d0afc891a950616b0d2e91fc8e48ebb60483de5a7326">+72/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>constants.py</strong><dd><code>Define global constants for SDK.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/constants.py

<li>Defined global constants for environment variables and local cache.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-7fff7186903cc72ead44eb4a912bb13fc4e90c993a7f342570afb55ddd90d7d8">+18/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>exceptions.py</strong><dd><code>Define base exception class for SDK.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/exceptions.py

- Defined base exception class for SDK errors.

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-b1d8af43031a39f00e1a6dfec0c947e296417d412b46db1c44f276b33845e3aa">+26/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>base.py</strong><dd><code>Define local storage class and methods.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/storage/base.py

<li>Defined <code>LocalStorage</code> class for local storage operations.<br> <li> Added methods for JSON serialization and file storage.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-8490641475214a57fb08b938b2017adf851f3b610bc43dbf5a92abd7b3e8cce2">+79/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>user.py</strong><dd><code>Define user data storage class.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/storage/user.py

- Defined `UserData` class for local user data storage.

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-85b3c2054e9fac70d939b322ab30b74393060ebc9f30dd9ad6558cfe0e9b975d">+18/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Add module docstring for storage helpers.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
composio/storage/__init__.py

- Added module docstring for storage helpers.

</details>
    

  </td>
  <td><a href="https://github.com/SamparkAI/composio/pull/53/files#diff-c9d1702f408d80e673d0c9713c8e87306c6593c291f6a32ab66d248c8e7c4751">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Review by utkarsh-dixit — CHANGES_REQUESTED** [review]

File: composio/client/endpoints.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        endpoint = endpoint or ""
+        if not endpoint.startswith("/"):
+            endpoint = f"/{endpoint}"
+        self.endpoint = endpoint
+
+    def __str___(self) -> str:
+        """String representation of the endpoint."""
+        return self.endpoint
+
+    __repr__ = __str___
```
Comment (kaavee315) [inline composio/client/endpoints.py:38]: As mentioned by Codium, seems like a bug

File: plugins/openai/composio_openai/toolset.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+
+class ComposioToolSet(BaseComposioToolSet):
+    """
+    Composio toolset for OpenAI framework.
+
+    Example:
+    ```python
+        import dotenv
+        from composio_openai import App, ComposioToolset
```
Comment (sohamganatra) [inline plugins/openai/composio_openai/toolset.py:31]: This is incorrect example and does not work. ComposioToolSet 
[S] is capital.

File: plugins/openai/composio_openai/toolset.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+        # Initialize tools.
+        openai_client = OpenAI()
+        composio_tools = ComposioToolset()
+
+        # Define task.
+        task = "Star a repo SamparkAI/composio_sdk on GitHub"
+
+        # Get GitHub tools that are pre-configured
+        actions = composio_toolset.get_tools(tools=[App.GITHUB])
```
Comment (sohamganatra) [inline plugins/openai/composio_openai/toolset.py:46]: Seems incorrect. It is apps=[App.Github] and not tools. May be we rename the function to get_apps??

**Comment by kaavee315** [comment]
**Viraj** Install pullflow and github auth

**Pull request merged**